From 9c7f42c9ecb1e6784bb53064d2dd7ced82b9d6b6 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 27 Jun 2024 09:53:25 +0200
Subject: [PATCH 0959/1695] drivers: stm32_firewall: remove old proprietary
 firewall framework

Since a generic firewall framework is available and every stm32 drivers
is plugged on it, remove this old proprietary firewall framework.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2f831810c8383b509451136ae616b7832ccb8174 drivers: stm32_firewall: remove old proprietary firewall framework]
Change-Id: If49fc0bbba8b6c29e9e21c5675334221fed99bc1
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/390162
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     |   1 -
 core/drivers/firewall/stm32_firewall.c        | 274 -----------------
 core/drivers/firewall/sub.mk                  |   1 -
 core/include/drivers/stm32_firewall.h         | 275 ------------------
 4 files changed, 551 deletions(-)
 delete mode 100644 core/drivers/firewall/stm32_firewall.c
 delete mode 100644 core/include/drivers/stm32_firewall.h

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index 19a2a9e6c..84c9425e4 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -14,7 +14,6 @@
 #ifdef CFG_STM32_CPU_OPP
 #include <drivers/stm32_cpu_opp.h>
 #endif
-#include <drivers/stm32_firewall.h>
 #include <drivers/stm32_remoteproc.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
diff --git a/core/drivers/firewall/stm32_firewall.c b/core/drivers/firewall/stm32_firewall.c
deleted file mode 100644
index 4b3c90f9f..000000000
--- a/core/drivers/firewall/stm32_firewall.c
+++ /dev/null
@@ -1,274 +0,0 @@
-// SPDX-License-Identifier: BSD-2-Clause
-/*
- * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
- */
-
-#include <assert.h>
-#include <drivers/stm32_firewall.h>
-#include <kernel/boot.h>
-#include <kernel/dt_driver.h>
-#include <kernel/spinlock.h>
-#include <libfdt.h>
-#include <util.h>
-
-#define MAX_COMPAT_LENGTH	30
-
-static LIST_HEAD(fw_dev_list_head, stm32_firewall_device) fw_dev_list =
-	LIST_HEAD_INITIALIZER(fw_dev_list_head);
-
-static unsigned int list_lock = SPINLOCK_UNLOCK;
-
-static const char firewall_compatible_exceptions[][MAX_COMPAT_LENGTH] = {
-	"st,stm32mp15-i2c-non-secure",
-	"st,stm32-ltdc"
-};
-
-static struct stm32_firewall_device *find_device(paddr_t base, size_t size,
-						 unsigned int *idx)
-{
-	struct stm32_firewall_device *fdev = NULL;
-	unsigned int i = 0;
-
-	assert(idx);
-
-	LIST_FOREACH(fdev, &fw_dev_list, dev_list) {
-		for (i = 0; i < fdev->compat->compat_size; i++) {
-			struct stm32_firewall_reg *reg = fdev->compat->reg + i;
-
-			if (!reg->addr && !reg->size)
-				continue;
-
-			if (!reg->size) {
-				if (base != reg->addr)
-					continue;
-
-				if (size)
-					DMSG("IOMEM %#"PRIxPA": non-null size",
-					     base);
-
-				*idx = i;
-				return fdev;
-			}
-
-			if (reg->size && base >= reg->addr &&
-			    base + size - 1 <= reg->addr + reg->size - 1) {
-				*idx = i;
-				return fdev;
-			}
-		}
-	}
-
-	return NULL;
-}
-
-TEE_Result stm32_firewall_check_access(paddr_t base, size_t size,
-				       const struct stm32_firewall_cfg *cfg)
-{
-	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
-	struct stm32_firewall_device *fdev = NULL;
-	uint32_t exceptions = 0;
-	unsigned int index = 0;
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	fdev = find_device(base, size, &index);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	if (fdev && fdev->ops->has_access)
-		res = fdev->ops->has_access(fdev, index, base, size, cfg);
-
-	return res;
-}
-
-TEE_Result stm32_firewall_set_config(paddr_t base, size_t size,
-				     const struct stm32_firewall_cfg *cfg)
-{
-	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
-	struct stm32_firewall_device *fdev = NULL;
-	uint32_t exceptions = 0;
-	unsigned int index = 0;
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	fdev = find_device(base, size, &index);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	if (fdev && fdev->ops->configure_access)
-		res = fdev->ops->configure_access(fdev, index, base, size, cfg);
-
-	return res;
-}
-
-TEE_Result stm32_firewall_get_data_config(paddr_t base, size_t size,
-					  const void *fdt, int node,
-					  int *ndata, void **data_cfg)
-{
-	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
-	struct stm32_firewall_device *fdev = NULL;
-	uint32_t exceptions = 0;
-	unsigned int index = 0;
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	fdev = find_device(base, size, &index);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	if (fdev && fdev->ops->get_data_config)
-		res = fdev->ops->get_data_config(fdt, node, ndata, data_cfg);
-
-	return res;
-}
-
-TEE_Result stm32_firewall_set_data_config(paddr_t base, size_t size,
-					  int ndata, void *data_cfg)
-{
-	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
-	struct stm32_firewall_device *fdev = NULL;
-	uint32_t exceptions = 0;
-	unsigned int index = 0;
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	fdev = find_device(base, size, &index);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	if (fdev && fdev->ops->set_data_config)
-		res = fdev->ops->set_data_config(ndata, data_cfg);
-
-	return res;
-}
-
-TEE_Result stm32_firewall_put_data_config(paddr_t base, size_t size,
-					  int ndata, void *data_cfg)
-{
-	TEE_Result res = TEE_ERROR_ITEM_NOT_FOUND;
-	struct stm32_firewall_device *fdev = NULL;
-	uint32_t exceptions = 0;
-	unsigned int index = 0;
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	fdev = find_device(base, size, &index);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	if (fdev && fdev->ops->put_data_config)
-		res = fdev->ops->put_data_config(ndata, data_cfg);
-
-	return res;
-}
-
-struct stm32_firewall_device *stm32_firewall_dev_alloc(void)
-{
-	return calloc(1, sizeof(struct stm32_firewall_device));
-}
-
-void stm32_firewall_dev_free(struct stm32_firewall_device *fdev)
-{
-	uint32_t exceptions = 0;
-
-	assert(fdev);
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	LIST_REMOVE(fdev, dev_list);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	free(fdev);
-}
-
-TEE_Result stm32_firewall_dev_register(struct stm32_firewall_device *fdev)
-{
-	uint32_t exceptions = 0;
-
-	assert(fdev && fdev->name && fdev->ops);
-
-	exceptions = cpu_spin_lock_xsave(&list_lock);
-	LIST_INSERT_HEAD(&fw_dev_list, fdev, dev_list);
-	cpu_spin_unlock_xrestore(&list_lock, exceptions);
-
-	return TEE_SUCCESS;
-}
-
-static TEE_Result
-stm32_firewall_bus_dt_register(struct stm32_firewall_device *fdev,
-			       const void *fdt, int node)
-{
-	const char __maybe_unused *name = NULL;
-	int idx = 0;
-	int len = 0;
-	int count = 0;
-	unsigned int id = 0;
-	const char *compat = NULL;
-	paddr_t pbase = 0;
-
-	/* Check accesses */
-	const struct stm32_firewall_cfg sec_cfg[] = {
-		{ FWLL_SEC_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
-	};
-
-	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
-		return TEE_SUCCESS;
-
-	count = fdt_stringlist_count(fdt, node, "compatible");
-	if (count < 0)
-		return TEE_SUCCESS;
-
-	pbase = fdt_reg_base_address(fdt, node);
-	if (pbase == DT_INFO_INVALID_REG)
-		return TEE_ERROR_GENERIC;
-
-	for (id = 0; id < fdev->compat->compat_size; id++)
-		if (fdev->compat->reg[id].addr == pbase)
-			break;
-
-	if (id == fdev->compat->compat_size)
-		return TEE_ERROR_ITEM_NOT_FOUND;
-
-	name = fdt_get_name(fdt, node, NULL);
-
-	compat = fdt_getprop(fdt, node, "compatible", &len);
-	assert(compat);
-
-	if (fdev->ops->has_access(fdev, id, pbase, 0, sec_cfg)) {
-		const struct stm32_firewall_cfg non_sec_cfg[] = {
-			{ FWLL_NSEC_PRIV_RW | FWLL_MASTER(0) },
-			{ }, /* Null terminated */
-		};
-		int len_array = ARRAY_SIZE(firewall_compatible_exceptions);
-		bool exception_found = false;
-
-		if (fdev->ops->has_access(fdev, id, pbase, 0, non_sec_cfg))
-			return TEE_SUCCESS;
-
-		for (int i = 0; i < len_array; i++) {
-			if (!strcmp(firewall_compatible_exceptions[i],
-				    compat))
-				exception_found = true;
-		}
-
-		if (!exception_found)
-			return TEE_SUCCESS;
-	}
-
-	for (idx = 0; idx < count; idx++) {
-		compat = fdt_stringlist_get(fdt, node, "compatible", idx, &len);
-		assert(compat && len > 0);
-
-		add_probe_node_by_compat(fdt, node, compat);
-	}
-
-	return TEE_SUCCESS;
-}
-
-TEE_Result stm32_firewall_bus_probe(struct stm32_firewall_device *fdev,
-				    const void *fdt, int node)
-{
-	TEE_Result res = TEE_ERROR_GENERIC;
-	int subnode = 0;
-
-	fdt_for_each_subnode(subnode, fdt, node) {
-		res = stm32_firewall_bus_dt_register(fdev, fdt, subnode);
-		if (res) {
-			DMSG("Failed on node %s with %#"PRIx32,
-			     fdt_get_name(fdt, subnode, NULL), res);
-			panic();
-		}
-	}
-
-	return TEE_SUCCESS;
-}
diff --git a/core/drivers/firewall/sub.mk b/core/drivers/firewall/sub.mk
index 69ceac0ee..ea2930c07 100644
--- a/core/drivers/firewall/sub.mk
+++ b/core/drivers/firewall/sub.mk
@@ -1,4 +1,3 @@
-srcs-y += stm32_firewall.c
 srcs-$(CFG_STM32_ETZPC) += stm32_etzpc.c
 srcs-$(CFG_STM32_IAC) += stm32_iac.c
 srcs-$(CFG_STM32_RIF) += stm32_rif.c
diff --git a/core/include/drivers/stm32_firewall.h b/core/include/drivers/stm32_firewall.h
deleted file mode 100644
index f6f35e16c..000000000
--- a/core/include/drivers/stm32_firewall.h
+++ /dev/null
@@ -1,275 +0,0 @@
-/* SPDX-License-Identifier: BSD-2-Clause */
-/*
- * Copyright (c) 2021, STMicroelectronics - All Rights Reserved
- */
-
-#ifndef __DRIVERS_STM32_FIREWALL_H
-#define __DRIVERS_STM32_FIREWALL_H
-
-#include <kernel/dt.h>
-#include <kernel/pm.h>
-#include <stdbool.h>
-#include <stdint.h>
-#include <sys/queue.h>
-#include <tee_api_types.h>
-#include <types_ext.h>
-
-/*
- * Firewall configuration
- *
- * 32b configuration ID contains, from LSbits to MSbits:
- * - 8 bit access permissions bit flags (rw, u/p, s/ns)
- * - 8 bit attribute flags (locking)
- * - 16 bit platform specific ID of the access master
- */
-
-/* Master configuration bits */
-#define FWLL_MASTER_SHIFT	16
-#define FWLL_MASTER(x)		((x) << FWLL_MASTER_SHIFT)
-#define FWLL_MASTER_MASK	GENMASK_32(31, 16)
-
-/* Lock configuration */
-#define FWLL_CONF_LOCK		BIT(15)
-
-/* Access configuration */
-#define FWLL_CONF_MASK		GENMASK_32(7, 0)
-#define FWLL_NO_ACCESS		0
-#define FWLL_SEC_READ		BIT(0)
-#define FWLL_SEC_WRITE		BIT(1)
-#define FWLL_SEC_PRIV		BIT(2)
-#define FWLL_SEC_UNPRIV		BIT(3)
-#define FWLL_NSEC_READ		BIT(4)
-#define FWLL_NSEC_WRITE		BIT(5)
-#define FWLL_NSEC_PRIV		BIT(6)
-#define FWLL_NSEC_UNPRIV	BIT(7)
-
-#define FWLL_SEC_PRIV_RW	(FWLL_SEC_READ | \
-				 FWLL_SEC_WRITE | \
-				 FWLL_SEC_PRIV)
-
-#define FWLL_SEC_UNPRIV_RW	(FWLL_SEC_READ | \
-				 FWLL_SEC_WRITE | \
-				 FWLL_SEC_UNPRIV)
-
-#define FWLL_SEC_RW		(FWLL_SEC_READ | \
-				 FWLL_SEC_WRITE | \
-				 FWLL_SEC_PRIV | \
-				 FWLL_SEC_UNPRIV)
-
-#define FWLL_NSEC_PRIV_RW	(FWLL_NSEC_READ | \
-				 FWLL_NSEC_WRITE | \
-				 FWLL_NSEC_PRIV)
-
-#define FWLL_NSEC_UNPRIV_RW	(FWLL_NSEC_READ | \
-				 FWLL_NSEC_WRITE | \
-				 FWLL_NSEC_UNPRIV)
-
-#define FWLL_NSEC_RW		(FWLL_NSEC_READ | \
-				 FWLL_NSEC_WRITE | \
-				 FWLL_NSEC_PRIV | \
-				 FWLL_NSEC_UNPRIV)
-
-#define FWLL_FULL_ACCESS	(FWLL_SEC_RW | \
-				 FWLL_NSEC_RW)
-
-struct stm32_firewall_device;
-
-/**
- * @brief struct stm32_firewall_reg - define the firewall device management
- *
- * @addr : controlled physical address.
- * @size : length of the controlled area.
- */
-struct stm32_firewall_reg {
-	paddr_t addr;
-	size_t size;
-};
-
-/**
- * @brief struct stm32_firewall_compat - Device tree compatibility table
- *
- * @reg: stm32_firewall_reg structure.
- * @compat_size: Array size of the stm32_firewall_reg structure.
- */
-struct stm32_firewall_compat {
-	struct stm32_firewall_reg *reg;
-	size_t compat_size;
-};
-
-/**
- * @brief struct stm32_firewall_cfg - Firewall configuration structure
- *
- * @access: Bit flags for permission, FWLL_[N]SEC_*
- */
-struct stm32_firewall_cfg {
-	uint32_t access;
-};
-
-/**
- * @brief struct stm32_firewall_ops
- *
- * @has_access: Check the access regarding the requested configuration.
- * @configure_access: Configure access based on requested configuration.
- * @request_access: Lock access for the requester.
- * @release_access: Unlock access of the requester.
- */
-struct stm32_firewall_ops {
-	TEE_Result (*has_access)(struct stm32_firewall_device *fdev,
-				 unsigned int id,
-				 paddr_t addr, size_t sz,
-				 const struct stm32_firewall_cfg *cfg);
-	TEE_Result (*configure_access)(struct stm32_firewall_device *fdev,
-				       unsigned int id,
-				       paddr_t addr, size_t sz,
-				       const struct stm32_firewall_cfg *cfg);
-	TEE_Result (*request_access)(struct stm32_firewall_device *fdev,
-				     unsigned int id,
-				     paddr_t addr, size_t sz,
-				     const struct stm32_firewall_cfg *cfg);
-	TEE_Result (*release_access)(struct stm32_firewall_device *fdev,
-				     unsigned int id,
-				     paddr_t addr, size_t sz,
-				     const struct stm32_firewall_cfg *cfg);
-	TEE_Result (*get_data_config)(const void *fdt, int node,
-				      int *ndata, void **data_cfg);
-	TEE_Result (*set_data_config)(int ndata, void *data_cfg);
-	TEE_Result (*put_data_config)(int ndata, void *data_cfg);
-};
-
-/**
- * @brief struct stm32_firewall_device
- *
- * @name: device name.
- * @ops: Operation table of the firewall.
- * @compat: List of the devices controlled by the firewall instance
- * @priv: Optional private data supplied by driver.
- * @dev_list: Firewall instance reference registered firewall list
- */
-struct stm32_firewall_device {
-	const char *name;
-	const struct stm32_firewall_ops *ops;
-	const struct stm32_firewall_compat *compat;
-	void *priv;
-
-	LIST_ENTRY(stm32_firewall_device) dev_list;
-};
-
-/* API for provider */
-
-/**
- * @brief stm32_firewall_priv - Retrieve firewall private data
- *
- * @fdev: stm32_firewall_device struct to be freed.
- *
- * Return firewall private data
- */
-static inline void *stm32_firewall_priv(struct stm32_firewall_device *fdev)
-{
-	return fdev->priv;
-}
-
-/**
- * @brief stm32_firewall_check_access - Check the configuration access
- * regarding a defined range.
- *
- * @base: Physical address
- * @size: Size to be checked
- * @cfg: Tested configuration
- *
- * Returns TEE_SUCCESS if access is permitted, or an error code otherwise
- */
-TEE_Result stm32_firewall_check_access(paddr_t base, size_t size,
-				       const struct stm32_firewall_cfg *cfg);
-
-/**
- * @brief stm32_firewall_set_config - Define the configuration access
- * regarding a defined range.
- *
- * @base: Physical address
- * @size: Size to be checked
- * @cfg: Tested configuration
- *
- * Returns a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_set_config(paddr_t base, size_t size,
-				     const struct stm32_firewall_cfg *cfg);
-
-/**
- * @brief stm32_firewall_get_data_config - Get the configuration access
- * from the device tree.
- *
- * @base: Physical address.
- * @size: Size to be checked.
- * @fdt: Device tree address.
- * @node: Bus node.
- * @ndata: Number of data configuration.
- * @data_cfg: Data configuration.
- *
- * Returns a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_get_data_config(paddr_t base, size_t size,
-					  const void *fdt, int node,
-					  int *ndata, void **data_cfg);
-
-/**
- * @brief stm32_firewall_set_data_config - Set the configuration access.
- *
- * @base: Physical address.
- * @size: Size to be checked.
- * @ndata: Number of data configuration.
- * @data_cfg: Data configuration.
- *
- * Returns a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_set_data_config(paddr_t base, size_t size,
-					  int ndata, void *data_cfg);
-
-/**
- * @brief stm32_firewall_put_data_config - Put the configuration access.
- *
- * @base: Physical address.
- * @size: Size to be checked.
- * @ndata: Number of data configuration.
- * @data_cfg: Data configuration.
- *
- * Returns a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_put_data_config(paddr_t base, size_t size,
-					  int ndata, void *data_cfg);
-
-/**
- * @brief stm32_firewall_dev_alloc - Allocate a new firewall device.
- *
- * Returns a stm32_firewall_device struct or NULL if allocation failed.
- */
-struct stm32_firewall_device *stm32_firewall_dev_alloc(void);
-
-/**
- * @brief stm32_firewall_dev_free - Free a firewall device.
- *
- * @fdev: stm32_firewall_device struct to be freed.
- */
-void stm32_firewall_dev_free(struct stm32_firewall_device *fdev);
-
-/**
- * @brief stm32_firewall_dev_register - Register a firewall device
- * list.
- *
- * @fdev: stm32_firewall_device struct to be added.
- *
- * Returns a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_dev_register(struct stm32_firewall_device *fdev);
-
-/**
- * @brief stm32_firewall_bus_probe - Probe a firewall bus driver child.
- *
- * @fdev: firewall_device struct to be probed.
- * @fdt: device tree address.
- * @node: firewall bus node.
- *
- * Return a TEE_Result compliant value.
- */
-TEE_Result stm32_firewall_bus_probe(struct stm32_firewall_device *fdev,
-				    const void *fdt, int node);
-#endif /* __DRIVERS_STM32_FIREWALL_H */
-- 
2.47.2

